src/shapes/Group.ts

```diff
@@ -483,7 +483,18 @@ export class Group extends createCollectionMixin(FabricObject<GroupEvents>) {
   drawObject(ctx: CanvasRenderingContext2D) {
     this._renderBackground(ctx);
     for (let i = 0; i < this._objects.length; i++) {
-      this._objects[i].render(ctx);
+      // TODO: handle rendering edge case somehow
+      if (
+        this.canvas?.preserveObjectStacking &&
+        this._objects[i].group !== this
+      ) {
+        ctx.save();
+        ctx.transform(...invertTransform(this.calcTransformMatrix()));
+        this._objects[i].render(ctx);
+        ctx.restore();
```

Comment:
this is the edge case handling, this is exactly how you do it.
The only improvement would be that instead of going back with an invert and a calc transform, you go straight to viewport transform using ctx.setTransform(this.canvas.viewportTransform).
I will add it in a follow up

Reply:
I think this logic should be handled by active selection.
We should discuss this, there are a lot of implications.